README.md
# ggirl - make ggplots in real life <img src='man/figures/logo.jpg' align="right" height="138.5" style="margin:10px;" />

_need support? Email [ggirl@jnolis.com](mailto:ggirl@jnolis.com)_

This package is a platform for taking ggplot2 objects and getting real life versions of them. Tired of saving your plots with the plain 'ol `ggsave()` function? Try this package instead!

The mediums in this package that you can choose from will change over time. Currently available are [ggwatercolor()](#ggwatercolor), [ggartprint()](#ggartprint), and [ggpostcard()](#ggpostcard). This package can only mail physical items to addresses in the United States.

## ggpostcard

`ggpostcard()` will take your ggplot2 object and _**will mail a postcard with it to the address of your choice!**_ Great for friends and colleagues (or maybe holiday cards??). You can specify a single address or many addresses to mail to, and you can customize the message on the back. If you send postcards to many people, you can customize the backs for each recipient if you so chose. _Each postcard costs $2.50 to send._

ORIGINAL GGPLOT2 PLOT             |  A REAL LIFE POSTCARD
:-------------------------:|:-------------------------:
![ggplot2 plot](man/figures/postcard-start.png)  | ![the resulting postcard](man/figures/postcard-end.jpg)


### How to make a postcard

First, install the package with:

```r
# install.packages("remotes") # if you don't already have it
remotes::install_github("jnolis/ggirl")
```

Then create a plot you like:

```r
library(ggplot2)
plot <- ggplot(data.frame(x=1:10, y=runif(10)),aes(x=x,y=y)) + geom_line() + geom_point()
```

Now let's ship it! Specify your email address (for order updates), address to mail to, and the message to put on the back:

```r
library(ggirl)
contact_email <- "fakeemailforreal@gmail.com"

send_address_1 <- address(name = "Fake Personname", address_line_1 = "250 North Ave",
                          city = "Boston", state = "MA",
                          postal_code = "22222", country = "US")

message_1 <- "This plot made me think of you!"

ggpostcard(plot, contact_email, messages = message_1, send_addresses = send_address_1)
```

_(You can include multiple recipients too and customize the return address--check out the function help for more details.)_

This will pop up a web page showing you what the order will look like. If both the front picture and back info look good to you click the button to make the payment and submit the order.

![The screen to review the order](man/figures/postcard-order-screen.png)

This will bring you to Stripe to finish the purchase.

![The screen to make the purchase](man/figures/postcard-payment-screen.png)

And you're done! Postcards will arrive in around 1-2 weeks.

### Examples for if you can't think of a cool postcard plot 

Included in ggirl::ggpostcard functionality are a few plots that are ready for you to try!

#### Sunrise/sunset plot

You can use `ggirl::ggpostcard_example_sunrise()` to make a postcard of sunrise and sunset locations for a city. It defaults to Seattle, WA, but you can use any location (check the documentation for requirements). It takes a minute or so to query the [sunrise/sunset api](https://sunrise-sunset.org/api):

<img src="man/figures/sunrise-example.png" alt-text="Example sunrise/sunset postcard" width="60%">

```r
library(ggirl)
contact_email <- "fakeemailforreal@gmail.com"
send_addresses <- address(name = "Fake Personname", address_line_1 = "250 North Ave",
                          city = "Boston", state = "MA",
                          postal_code = "22222", country = "US")
messages <- "Look at this cool plot I found!"

plot <- ggpostcard_example_sunrise()
ggpostcard(plot = plot,
           contact_email = contact_email,
           send_addresses = send_addresses, messages = messages)
```

#### ContouR plot

You can use `ggirl::ggpostcard_example_contouR()` to make a postcard of the generative art of [@Ijeamakaanyene](https://github.com/Ijeamakaanyene). Check the help function on how to change the colors of the plot.

<img src="man/figures/contouR-example.png" alt-text="Example contouR postcard" width="60%">

```r
library(ggirl)
contact_email <- "fakeemailforreal@gmail.com"
send_addresses <- address(name = "Fake Personname", address_line_1 = "250 North Ave",
                          city = "Boston", state = "MA",
                          postal_code = "22222", country = "US")
messages <- "Look at this cool plot I found!"

plot <- ggpostcard_example_contouR()
ggpostcard(plot = plot,
           contact_email = contact_email,
           send_addresses = send_addresses, messages = messages)
```

#### rstereogram plot

You can use `ggirl::ggpostcard_example_rstereogram()` to make a postcard of a stereogram of an image using the rstereogram package by [@ryantimpe](https://github.com/ryantimpe). It defaults to the R logo but you can pass it any PNG file. Check the documentation for best practices around that, and if you're having trouble seeing the image you may need to zoom in on your monitor.

<img src="man/figures/rstereogram-example.png" alt-text="Example rstereogram postcard" width="60%">

```r
library(ggirl)
contact_email <- "fakeemailforreal@gmail.com"
send_addresses <- address(name = "Fake Personname", address_line_1 = "250 North Ave",
                          city = "Boston", state = "MA",
                          postal_code = "22222", country = "US")
messages <- "Look at this cool plot I found!"

plot <- ggpostcard_example_rstereogram()
ggpostcard(plot = plot,
           contact_email = contact_email,
           send_addresses = send_addresses, messages = messages)
```

## ggartprint

![Art prints made by ggartprint](man/figures/artprint-examples.jpg)
_From left to right: [ggartprint_example_map](#open-street-map-plot) plot, [Palmer Penguin](https://github.com/allisonhorst/palmerpenguins) data, and ggplot2 art by [@ijeamakaanyene](https://github.com/Ijeamakaanyene)._

With `ggartprint()` you can order a beautiful art print to be mailed to you! The prints are on archival photo-rag paper and come in a number of sizes from "cute and small" to "large and centerpiece of a room"

### How to make an art print

First, install the package with:

```r
# install.packages("remotes") # if you don't already have it
remotes::install_github("jnolis/ggirl")
```

Then create a plot you like:

```r
library(ggplot2)
plot <- ggplot(data.frame(x=1:10, y=runif(10)),aes(x=x,y=y)) +
  geom_line() + geom_point() + theme_gray(48)
```

Now let's order the prints! You'll need to specify your email address and the physical address to mail to. You'll also need to specify the size (ex: "11x14"), the orientation (ex: "portrait"), and quantity (typically 1).

```r
library(ggirl)
delivery_address <- address(name = "Fake person", address_line_1 = "101 12th st",
address_line_2 = "Apt 17", city = "Seattle", state = "WA",
  postal_code = "98102", country = "US")
contact_email = "fakeemail275@gmail.com"
ggartprint(plot, size="11x14", orientation = "landscape", quantity = 1,
           contact_email = contact_email, address = delivery_address)
```

This will open a web page to preview the art print. if you like how it looks you can submit the order and pay via Stripe.

![The screen to review the order](man/figures/artprint-order-screen.png)

And you're done! Art prints will arrive in around 2-3 weeks.

### Examples for if you can't think of a cool art print plot 

#### Open Street Map plot

Try making a map! With `ggartprint_example_map()` you can make a map that would cost you twice as much from a non-R based tool. This uses the `osmdata` package to get data from Open Street Maps, and is based on a tutorial by [Joshua McCrain](http://joshuamccrain.com/tutorials/maps/streets_tutorial.html).

```r
library(ggirl)
contact_email <- "fakeemailforreal@gmail.com"
delivery_address <- address(name = "Fake Personname", address_line_1 = "250 North Ave",
                          city = "Boston", state = "MA",
                          postal_code = "22222", country = "US")
plot <- ggartprint_example_map("Seattle", "Washington")
# use the background option to avoid white boundaries
ggartprint(plot, background = "#FEFDF7", size = "11x14", orientation = "portrait",
                                contact_email = contact_email,
                                address = delivery_address))
```

![The screen to review the order](man/figures/map-example-order-screen.png)

## ggwatercolor

ORIGINAL GGPLOT2 PLOT             |  A REAL LIFE WATERCOLOR
:-------------------------:|:-------------------------:
![ggplot2 plot](man/figures/watercolor-start-sm.png)  | ![the resulting watercolor](man/figures/watercolor-end.jpg)


With `ggwatercolor()` you have a hand-painted watercolor painting of your ggplot2. These paintings are done
on heavy cold-pressed 8"x10" watercolor paper with lightfast paints, meaning your beautiful watercolor will withstand
the test of time.

### How to request a watercolor commission

First, install the package with:

```r
# install.packages("remotes") # if you don't already have it
remotes::install_github("jnolis/ggirl")
```

Then create a plot you like:

```r
library(ggplot2)
plot <- ggplot(data.frame(x=1:10, y=runif(10)),aes(x=x,y=y)) +
  geom_line() + geom_point() + theme_gray(48)
```

Then to request the commission! The `ggwatercolor()` function can be called to do so. Just specify the plot,
the orientation, a contact email, and the address to mail the commission to.

```r
library(ggirl)
delivery_address <- address(name = "Fake person", address_line_1 = "101 12th st",
address_line_2 = "Apt 17", city = "Seattle", state = "WA",
  postal_code = "98102", country = "US")
contact_email = "fakeemail275@gmail.com"
ggartprint(plot, orientation = "landscape", contact_email = contact_email, address = delivery_address)
```

This will open a web page to preview the watercolor. If you like it, you can request the commission, and in a
few days you'll get an email with details about the commission and request for payment.

![The screen to review the order](man/figures/watercolor-order-screen.png)

The email will include the price, between \$100-\$175 depending on the complexity of the picture, and a link to pay. If the plot is complex it may also need to be simplified by the artist. After payment, the commission should be complete in up to 4 weeks.


## Get involved

If you think this package is interesting you can help in multiple ways! Maybe you have an R package that could call one of these functions! Maybe you could come up with a new form of fulfillment, like making plots out of clay! Email ggirl@jnolis.com to discuss it.

## Acknowledgments

Thanks to:

* [@nolistic](https://github.com/nolistic) and [@robinsones](https://github.com/robinsones) for helping design the product.
* The ggpostcard beta testers: [@cxinya](https://github.com/cxinya), [@delabj](https://github.com/delabj), [@thisisnic](https://github.com/thisisnic), [@TheCoachEdwards](https://github.com/TheCoachEdwards), [@ryantimpe](https://github.com/ryantimpe), [@robinsones](https://github.com/robinsones), [@mcsiple](https://github.com/mcsiple), and [@cathblatter](https://github.com/cathblatter).
* [@ColinFay](https://github.com/ColinFay) for the package [{brochure}](https://github.com/ColinFay/brochure) which powers the Shiny server doing the back-end work.
* [@Ijeamakaanyene](https://github.com/Ijeamakaanyene) for the use of the contouR example and beta testing ggartprint.
* [@ryantimpe](https://github.com/ryantimpe) for the use of the rstereogram example.
* [sunrise-sunset.org](https://sunrise-sunset.org/) for the API powering the sunrise example.
* [Joshua McCrain](http://joshuamccrain.com/tutorials/maps/streets_tutorial.html) for the tutorial used for the map example.
* [The R Foundation](https://www.r-project.org/logo/) for the R logo in the ggpostcard_example_rstereogram().


NEWS.md
# Version 3.1.5 (2022-06-07)
Updated art print prices to reflect increased costs

# Version 3.1.4 (2022-10-21)
Improved the sunrise postcard example after being motivated by #normconf

# Version 3.1.3 (2022-10-12)
Backend changes:improves error handling if package cannot connect to server

# Version 3.1.2 (2022-08-14)
Backend changes

# Version 3.1.1 (2022-07-15)
Fixed a bug where the first request would fail if the server has been idle for a while

# Version 3.1.0 (2022-06-29)
Backend changes

# Version 3.0.1 (2021-12-08)
Minor bug fix with map example

# Version 3.0.0 (2021-08-10)
Now includes ggwatercolor() for commissioning hand-painted watercolor art of your ggplot

# Version 2.0.0 (2021-06-07)
Now includes ggartprint() for making art prints for your walls

# Version 1.0.1 (2021-04-15)
Improved max character computation by accounting for newlines

# Version 1.0.0 (2021-04-12)
Initial launch with ggirl::ggpostcard()


R/ggirl.R
#' Create an address object
#'
#' This function takes string inputs and converts them into an address object that can be used to send irl art (or as a return address).
#'
#' @param name The name for the address
#' @param address_line_1 The first line of the address
#' @param address_line_2 (Optional) A second address line, such as an apartment number.
#' @param city the city
#' @param state (Optional) The state to send to
#' @param postal_code The postal code (ZIP code in the US)
#' @param country The 2-character [ISO-1366 code](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) for the country. Non-US shipping is experimental!
#'
#' @examples
#' send_address <- address(name = "RStudio", address_line_1 = "250 Northern Ave",
#'   city = "Boston", state = "MA", postal_code = "02210", country = "US")
#'
#' @export
address <- function(name,
                    address_line_1,
                    address_line_2 = NULL,
                    city,
                    state = NULL,
                    postal_code,
                    country){
  address_set <- list(name = name,
                      address_line_1 = address_line_1,
                      address_line_2 = address_line_2,
                      city = city,
                      state = state,
                      postal_code = postal_code,
                      country = country)
  # Check country is valid
  if (!is.character(country) || nchar(country) != 2){
    stop("Country must be a 2-character ISO-1366 code (https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes)")
  }

  structure(address_set, class="ggirl_address")
}

#' Get the server URL
#'
#' This function gets the most current server URL by using a fixed lookup URL
get_server_url <- function(){
  server_url <- getOption("ggirl_server_url", "https://skyetetra.shinyapps.io/ggirl-server")

  tryCatch({
    httr::RETRY(
      verb = "GET",
      url = server_url,
      times = 2,
      pause_min = 5,
      pause_cap = 5,
      quiet = TRUE,
      httr::timeout(5)
    )
  }, error = function(e){
    stop("ggirl server is not connecting--try updating the ggirl package or email ggirl@jnolis.com")
  })

  server_url
}


#' Get package version
#'
#' @return the version of the package being used
get_version <- function(){
  version <- packageDescription("ggirl", fields = "Version")
  if(is.na(version)){
    version <- "0.0.0"
  }
  version
}


#' Upload data
#'
#' This function will upload the data to the server, then launch the page for it
upload_data_and_launch <- function(data, server_url, type){
  zz <- rawConnection(raw(0), "r+")
  on.exit({close(zz)}, add=TRUE)
  saveRDS(data, zz)
  seek(zz, 0)

  tryCatch({
    url <- paste0(server_url, "/upload")
    body <- rawConnectionValue(zz)
    response <- httr::POST(url = url,
               body = body,
               httr::content_type("application/octet-stream"),
               httr::timeout(5))
    if(!(response$status_code %in% c(200, 201, 202))){
      message("Waiting for the server to respond (may take up to 30 seconds)")
      response <- httr::RETRY(
        verb = "POST",
        url = url,
        body = body,
        pause_min = 3,
        times = 5,
        pause_cap = 5,
        quiet = TRUE,
        terminate_on = c(200, 201, 202),
        httr::content_type("application/octet-stream"),
        httr::timeout(5)
      )
    }
    if(!(response$status_code %in% c(200, 201, 202))){
      stop("Invalid response")
    }
    token <- httr::content(response, as="text", encoding="UTF-8")
    browseURL(paste0(server_url,"/", type, "?token=",token))
  }, error = function(e){
    stop("Plot upload failed--try updating the ggirl package or email ggirl@jnolis.com")
  })
}


R/ggwatercolor.R
watercolor_dpi <- 300
watercolor_width <- 10
watercolor_height <- 8

#' Preview your watercolor request
#'
#' This function takes a ggplot2 output and gives a preview of how plot will be requested to be painted.
#' While it's totally fine to just call ggirl::ggwatercolor to preview, this allows you to preview locally.
#'
#' The preview will appear in either the "Viewer" pane of RStudio or in your browser, depending on if RStudio is installed or not.
#' The preview includes a frame, but that will not be included with the print.
#'
#' @param plot the plot to use as an art print
#' @param orientation should the plot be landscape or portrait?
#' @param ... other options to pass to `ragg::agg_png()` when turning the plot into an image.
#' @seealso [ggwatercolor()] to request the watercolor commission
#' @examples
#' library(ggplot2)
#' library(ggirl)
#' plot <- ggplot(data.frame(x=1:10, y=runif(10)),aes(x=x,y=y))+geom_line()+geom_point()+theme_gray(48)
#' ggwatercolor_preview(plot, orientation = "landscape")
#' @export
ggwatercolor_preview <- function(plot, size, orientation, ...){
  temp_dir <- tempfile()
  dir.create(temp_dir)
  temp_plot_file <- file.path(temp_dir, "plot.png")
  temp_css_file <- file.path(temp_dir, "site.css")
  temp_html_file <- file.path(temp_dir, "index.html")
  mg <- ceiling(postcard_width_px*(safe_margin-cut_margin))
  css <- "body {margin: 0;}
.frame {
    background-color: #303030;
    box-shadow: 0 10px 7px -5px rgba(0, 0, 0, 0.3);
    padding: 1rem!important;
    margin: 1rem!important;
    display: inline-block;
}

.box-shadow {
    position: relative;
    text-align: center;
}

.box-shadow::after {
    box-shadow: 0px 0px 20px 0px rgba(0,0,0,0.5) inset;
    bottom: 0;
    content: '';
    display: block;
    left: 0;
    height: 100%;
    position: absolute;
    right: 0;
    top: 0;
    width: 100%;
}
.box-shadow img {
    max-width: 100%;
    width: auto;
    max-height: 90vh;
}

.watercolor {
    max-width: 100%;
    height: auto;
}"

  html <- '
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>ggartprint preview</title>
    <link rel="stylesheet" href="site.css">
  </head>
  <body>
  <div class="frame">
  <div class="box-shadow">
    <img src="plot.png" class = "watercolor">
  </div>
  </div>
  </body>
</html>
  '

  ggwatercolor_save(filename = temp_plot_file, plot = plot, orientation = orientation, ...)
  writeLines(css, temp_css_file)
  writeLines(html, temp_html_file)
  viewer <- getOption("viewer")
  if (!is.null(viewer))
    viewer(temp_html_file)
  else
    utils::browseURL(temp_html_file)
}

ggwatercolor_save <- function(filename, plot, orientation = c("landscape","portrait"), ...){
  orientation <- match.arg(orientation)

  if(orientation == "portrait"){
    width <- watercolor_height
    height <- watercolor_width
  } else if(orientation == "landscape") {
    width <- watercolor_width
    height <- watercolor_height
  } else {
    stop("invalid orientation")
  }

  old_dev <- grDevices::dev.cur()
  ragg::agg_png(
    filename,
    width = width,
    height = height,
    units = "in",
    res = watercolor_dpi,
    ...)

  on.exit(utils::capture.output({
    grDevices::dev.off()
    if (old_dev > 1) grDevices::dev.set(old_dev)
  }))

  grid::grid.draw(plot)

}


#' Request a watercolor commission of your ggplot!
#'
#' This function takes a ggplot2 output and request a handpainted watercolor painting of it!
#' Running this function will bring you to a webpage to confirm the request--a followup email will contain the details and how to pay.
#' _No painting will be made until after an email exchange and the payment sent._
#'
#' Watercolor paintings will be made 8"x10" on 140lb cold-press paper with professional light-safe paints.
#' Since these are painted by hand, they may not be exactly accurate to the original, and may be simplified
#' depending on the complexity of the original (which will be discussed by email).
#'
#' The paintings take up to 4 weeks to be delivered.
#'
#' @param plot the plot to use as an art print.
#' @param orientation should the plot be landscape or portrait?
#' @param contact_email email address to send order updates.
#' @param quantity the number of prints to order (defaults to 1).
#' @param address the physical address to mail the painting to. Use the [address()] function to format it.
#' @param ... other options to pass to `ragg::agg_png()` when turning the plot into an image for the front of the postcard.
#' @seealso [address()] to format an address for ggirl
#' @examples
#' library(ggplot2)
#' library(ggirl)
#' delivery_address <- address(name = "Fake person", address_line_1 = "101 12th st",
#'   address_line_2 = "Apt 17", city = "Seattle", state = "WA",
#'   postal_code = "98102", country = "US")
#' contact_email = "fakeemail275@gmail.com"
#' plot <- ggplot(data.frame(x=1:10, y=runif(10)),aes(x=x,y=y))+geom_line()+geom_point()+theme_gray(48)
#' ggwatercolor(plot, orientation = "landscape",
#'            contact_email = contact_email, address = delivery_address)
#' @export
ggwatercolor <- function(plot, orientation = c("landscape","portrait"), contact_email, address, ...){

  orientation <- match.arg(orientation)
  if(any(address$country != "US")){
    stop("Art prints only available for US addresses through package. Email ggirl@jnolis.com to price a custom order.")
  }

  version <- get_version()
  server_url <- get_server_url()

  temp_png <- tempfile(fileext = ".png")
  on.exit({file.remove(temp_png)}, add=TRUE)
  ggwatercolor_save(filename = temp_png, plot=plot, orientation = orientation, ...)
  raw_plot <- readBin(temp_png, "raw", file.info(temp_png)$size)

  data <- list(
    type = "watercolor",
    contact_email = contact_email,
    raw_plot = raw_plot,
    address = address,
    orientation = orientation,
    version = version
  )

  upload_data_and_launch(data, server_url, "watercolor")
}


R/ggpostcard_example_sunrise.R
get_sunrise_sunset_single_day <- function(coordinates, date, tz){
  response <- httr::GET(glue::glue("https://api.sunrise-sunset.org/json?lat={coordinates$lat}",
                                   "&lng={coordinates$long}&date={date}&formatted=0"))
  results <- httr::content(response)$results

  results[["day_length"]] <- NULL
  results <- lapply(results, function(x) as.POSIXct(x,format="%Y-%m-%dT%T",tz="UTC"))
  results <- lapply(results, function(x) lubridate::with_tz(x, tzone = tz))
  results
}

get_year_dates <- function(year = NULL){
  if(is.null(year)){
    year <- lubridate::year(Sys.time())
  }
  start_date <- as.Date(paste0(year,"-01-01")) - lubridate::days(3)
  end_date <- as.Date(paste0(year + 1,"-01-01")) + lubridate::days(3)
  seq.Date(start_date,end_date,by="day")
}


mem_get_sunrise_sunset_single_day <- memoise::memoise(get_sunrise_sunset_single_day)



get_sunrise_sunsets <- function(lat, long, tz, year, progress = TRUE){

  dates <- get_year_dates(year)

  message("Querying api: https://sunrise-sunset.org/api")
  if(progress){
    pb <- progress::progress_bar$new(format = "  calling sunrise/sunset api [:bar] :percent eta: :eta",
                                     total = length(dates))
  }

  if(exists("mem_get_sunrise_sunset_single_day")){
    get_fun <- mem_get_sunrise_sunset_single_day
  } else {
    get_fun <- get_sunrise_sunset_single_day
  }
  datetimes <- lapply(dates, function(d){
    pb$tick()
    get_fun(list(lat = lat, long = long), d, tz)
  })
  result <- do.call(rbind,lapply(datetimes,as.data.frame))
  result$date <- dates
  result
}


format_sunrise_info <- function(sunrise_info, year, tz){
  night_type_lookup <- data.frame(
    name =
      c("sunrise", "sunset",
        "civil_twilight_begin", "civil_twilight_end",
        "nautical_twilight_begin", "nautical_twilight_end",
        "astronomical_twilight_begin", "astronomical_twilight_end"),
    night_group =
      c("sunrise/sunset", "sunrise/sunset",
        "civil_twilight", "civil_twilight",
        "nautical_twilight", "nautical_twilight",
        "astronomical_twilight", "astronomical_twilight"),
    night_type =
      c("begin", "end",
        "begin", "end",
        "begin", "end",
        "begin", "end")
  )

  si_long <- tidyr::pivot_longer(sunrise_info, -date)
  y_origin <- as.POSIXct(as.Date(paste0(year,"-01-01")), tz="UTC")

  strip_date <- function(local_time,origin,tz){
    stripped <- y_origin +
      lubridate::seconds(as.numeric(difftime(local_time,
                                             as.POSIXct(paste0(lubridate::ymd(origin)," 00:00:00"),tz=tz),
                                             units="secs")))
    stripped <- lubridate::with_tz(stripped, tz = "UTC")
    stripped[stripped < as.Date("1972-01-01")] <- NA
    stripped
  }

  si_long <- dplyr::left_join(si_long,night_type_lookup, by = "name")

  si_long_begin <- dplyr::filter(si_long,night_type == "begin")
  si_long_noon <- dplyr::filter(si_long,is.na(night_type))
  si_long_end <- dplyr::filter(si_long,night_type == "end")

  si_long_begin_joined <-
    dplyr::inner_join(
      dplyr::select(si_long_begin, date, name, night_group, begin_value = value),
      dplyr::mutate(dplyr::select(si_long_end,
                                  date, night_group,
                                  end_value = value),
                    date = date + lubridate::days(1)),
      by = c("night_group","date"))

  si_long_end_joined <-
    dplyr::inner_join(
      dplyr::select(si_long_end, date, name, night_group, end_value = value),
      dplyr::mutate(dplyr::select(si_long_begin,
                                  date, night_group,
                                  begin_value = value),
                    date = date - lubridate::days(1)),
      by = c("night_group","date"))

  si_long_lines <- dplyr::mutate(si_long_noon,
                                 stripped_value = strip_date(value, date, tz))

  si_long_areas <- dplyr::bind_rows(si_long_begin_joined, si_long_end_joined)
  si_long_areas <- dplyr::mutate(si_long_areas,
                                 begin_stripped_value = strip_date(begin_value, date, tz),
                                 end_stripped_value = strip_date(end_value, date, tz))


  list(si_long_areas = si_long_areas, si_long_lines = si_long_lines, y_origin = y_origin, year = year)
}

make_sunrise_plot_simple <- function(formatted_sunrise_info, location_name){
  plot_order <- function(name){
    factor(name,
           levels =rev(c("astronomical_twilight_begin",
                         "nautical_twilight_begin",
                         "civil_twilight_begin",
                         "sunrise",
                         "solar_noon",
                         "astronomical_twilight_end",
                         "nautical_twilight_end",
                         "civil_twilight_end",
                         "sunset"
           )))
  }

  year <- formatted_sunrise_info$year
  y_origin <- formatted_sunrise_info$y_origin
  si_long_areas<- formatted_sunrise_info$si_long_areas
  si_long_lines <- formatted_sunrise_info$si_long_lines

  si_long_areas$plot_order <- plot_order(si_long_areas$name)

  light_color <- "#D0D0D8"
  dark_color <- "#181820"

  very_light_blue <- "#c5d2fa"
  light_blue <- "#95abf0"
  mid_blue <- "#7691e3"
  dark_blue <- "#4a65b6"
  very_dark_blue <- "#343e6b"
  black <- "#0A0D29"

  pretty_time <- function(times){
    paste0(as.numeric(format(times,format = "%I", tz = "UTC")),
           tolower(format(times,format = "%p", tz = "UTC")))
  }

  colors <- c(black, very_dark_blue, dark_blue, mid_blue, black, very_dark_blue,dark_blue, mid_blue)
  ggplot2::ggplot() +
    ggplot2::geom_ribbon(data = si_long_areas,
                         ggplot2::aes(x = date,
                                      ymin = begin_stripped_value,
                                      ymax = end_stripped_value,
                                      fill = plot_order)) +
    ggplot2::geom_line(data = si_long_lines,
                       ggplot2::aes(x=date,y=stripped_value),
                       color = dark_blue)+
    ggplot2::coord_cartesian(ylim = y_origin+lubridate::hours(c(0,24)),
                             xlim = c(as.Date(paste0(year,"-01-01")), as.Date(paste0(year+1,"-01-01")))) +
    ggplot2::scale_y_datetime(breaks = y_origin+lubridate::hours(c(0,6,12,18,24)),
                              expand = ggplot2::expansion(0,0),
                              labels = pretty_time)+
    ggplot2::scale_x_date(expand = ggplot2::expansion(0,0),
                          date_labels = "%b")+
    ggplot2::geom_hline(data=data.frame(yintercept = y_origin +
                                          lubridate::hours(c(0, 6, 12, 18, 24))),
                        ggplot2::aes(yintercept=yintercept), color=light_color, size=0.65, alpha=0.5)+
    ggplot2::geom_hline(data=data.frame(yintercept = y_origin +
                                          lubridate::hours(seq(0,24,3))),
                        ggplot2::aes(yintercept=yintercept), color=light_color, size=0.325, alpha=0.5)+
    ggplot2::geom_hline(data=data.frame(yintercept = y_origin +
                                          lubridate::hours(seq(-1,25,1))),
                        ggplot2::aes(yintercept=yintercept), color=light_color, size=0.125, alpha=0.5)+
    ggplot2::geom_vline(data=data.frame(xintercept = seq.Date(as.Date(y_origin), as.Date(y_origin + lubridate::days(366)), by="3 months")),
                        ggplot2::aes(xintercept=xintercept), color=light_color, size=0.625, alpha=0.5)+
    ggplot2::geom_vline(data=data.frame(xintercept = seq.Date(as.Date(y_origin), as.Date(y_origin + lubridate::days(366)), by="month")),
                        ggplot2::aes(xintercept=xintercept), color=light_color, size=0.125, alpha=0.5)+
    ggplot2::scale_fill_manual(values = rev(colors)) +
    ggplot2::theme_minimal(8) +
    ggplot2::theme(panel.background = ggplot2::element_rect(fill = very_light_blue, size = 0, color = NA),
                   panel.border = ggplot2::element_blank(),
                   plot.background = ggplot2::element_rect(fill = dark_color, size = 0, color = NA),
                   panel.grid =ggplot2:: element_blank(),
                   legend.title = ggplot2::element_blank(),
                   legend.text = ggplot2::element_text(color = light_color),
                   axis.title = ggplot2::element_blank(),
                   axis.text = ggplot2::element_text(size = ggplot2::rel(1.8), color = light_color),
                   axis.text.y = ggplot2::element_text(margin = ggplot2::margin(r = 5)),
                   axis.text.x = ggplot2::element_text(margin = ggplot2::margin(t = 5)),
                   plot.title = ggplot2::element_text(size = ggplot2::rel(4), color = light_color, margin = ggplot2::margin(b=3)),
                   plot.subtitle = ggplot2::element_text(size = ggplot2::rel(1.25), color = light_color,
                                                         margin = ggplot2::margin(b=15,t=0)),
                   plot.margin = ggplot2::margin(r = 30, t=30, l = 30, b=30),
                   plot.caption = ggplot2::element_text(size = ggplot2::rel(0.8), color = light_color),
                   plot.title.position = "plot",
                   plot.caption.position = "plot",
                   legend.position="none"
    ) +
    ggplot2::labs(title = location_name,
                  subtitle = paste0("Night | Astronomical Twilight | Nautical Twilight | Civil Twilight | Daylight"))

}

#' An example postcard with the sunrise and sunset times for a location
#'
#' This example makes a postcard with a pretty ggplot that gives you when sunrise
#' and sunset is over the course of the year.
#'
#' To get the longitude and latitude for your location, you can click a place in Google maps.
#' To get the time zone, you can use the \href{https://en.wikipedia.org/wiki/List_of_tz_database_time_zones}{Wikipedia list of time zone codes}
#'
#' This function uses an external API \href{https://sunrise-sunset.org/api}{https://sunrise-sunset.org/api}. While the package has some built-in caching,
#' _if you call the function with too many locations in a short period of time your IP will be temporarily blocked_.
#'
#' If all of the inputs are NULL, the function defaults to using Seattle, Washington.
#'
#' @param location_lat the latitude of the location
#' @param location_long the longitude of the location
#' @param location_tz a tz time zone string of the style taken by R functions like as.POSIXct
#' @param location_name string for the name to show for the location (ex: "Seattle, WA")
#' @return a ggplot2 plot to pass to ggpostcard
#' @family ggpostcard_examples
#' @examples
#' library(ggirl)
#' location_lat <- 47.6062
#' location_long <- -122.3321
#' location_tz <- "America/Los_Angeles"
#' location_name <- "Seattle, WA"
#' return_address <- address(name = "Jacqueline Nolis", address_line_1 = "111 North St",
#'                           city = "Seattle", state = "WA",
#'                           postal_code = "11111", country = "US")
#' contact_email <- "fakeemailforreal@gmail.com"
#' send_addresses <- address(name = "Fake Personname", address_line_1 = "250 North Ave",
#'                           city = "Boston", state = "MA",
#'                           postal_code = "22222", country = "US")
#' messages <- "Look at this cool plot I found!"
#' plot <- ggpostcard_example_sunrise(location_lat, location_long, location_tz, location_name)
#' ggpostcard(plot = plot,
#'   contact_email = contact_email, return_address = return_address,
#'   send_addresses = send_addresses, messages = messages)
#' @export
ggpostcard_example_sunrise <- function(location_lat = NULL, location_long = NULL, location_tz = NULL, location_name = NULL, year = NULL, ...){
  required_packages <- c("lubridate", "progress", "tidyr", "dplyr")
  packages_is_installed <- sapply(required_packages, function(x) requireNamespace(x, quietly = TRUE))

  if (any(!packages_is_installed)) {
    stop(paste0("This example requires you to install these packages: ",
                paste0(required_packages[!packages_is_installed], collapse = ", ")),
         call. = FALSE)
  }

  if(is.null(year)){
    year <- lubridate::year(Sys.Date())
  }

  if(is.null(location_lat) || is.null(location_long) || is.null(location_tz) || is.null(location_name)){
    stop("Must supply location_lat, location_long, location_tz, and location_name")
  }

  sunrise_info <- get_sunrise_sunsets(location_lat, location_long, location_tz, year)
  formatted_sunrise_info <- format_sunrise_info(sunrise_info, year, location_tz)
  plot <- make_sunrise_plot_simple(formatted_sunrise_info, location_name)
  plot
}
